turn the futures-unordered footgun into a status quo story (#172)
rust-lang/wg-async-foundations: src/vision/status_quo/barbara_battles_buffered_streams.md

## 🚧 Warning: Draft status 🚧

This is a draft "status quo" story submitted as part of the brainstorming period. It is derived from real-life experiences of actual Rust users and is meant to reflect some of the challenges that Async Rust programmers face today. 

If you would like to expand on this story, or adjust the answers to the FAQ, feel free to open a PR making edits (but keep in mind that, as they reflect peoples' experiences, status quo stories [cannot be wrong], only inaccurate). Alternatively, you may wish to [add your own status quo story][htvsq]!

## The story

### Mysterious timeouts

Barbara is working on her [YouBuy] server and is puzzling over a strange bug report. She is encountering users reporting that their browser connection is timing out when they connect to [YouBuy]. Based on the logs, she can see that they are timing out in the `do_select` function:

```rust
async fn do_select<T>(database: &Database, query: Query) -> Result<Vec<T>> {
    let conn = database.get_conn().await?;
    conn.select_query(query).await
}
```

> 

Reviewer: ```suggestion
This is surprising, because `do_select` doesn't do much - it does a database query to claim a work item from a queue, but isn't expected to handle a lot of data or hit extreme slowdown on the database side.
```

Some of the time, there is some kind of massive delay in between the `get_conn` method opening a connection and the call to `select_query`. But why? She has metrics that show that the CPU is largely idle, so it's not like the cores are all occupied.

She looks at the caller of `do_select`, which is a function `do_work`:

```rust
async fn do_work(database: &Database) {
    let work = do_select(database, FIND_WORK_QUERY)?;

>     stream::iter(work.into_iter())

Reviewer 2: This method can accept any `IntoIterator`.
```suggestion
    stream::iter(work)
```

        .map(|item| do_select(database, work_from_item(item)))
        .buffered(5)
        .for_each(|work_item| process_work_item(database, work_item))
        .await;
}

async fn process_work_item(...) { }
```

The `do_work` function is invoking `do_select` as part of a stream; it is buferring up a certain number of `do_select` instances and, for each one, invoking `process_work_item`. Everything seems to be in order, and she can see that calls to `process_work_item` are completing in the logs.

> Following a hunch, she adds more logging in and around the `process_work_item` function and waits a few days to accumulate new logs. She notices that shortly after each time out, there is always a log of a `process_work_item` call that takes at least 20 seconds. These calls are not related to the connections that time out, they are for other connections, but they always appear afterwards in time.

Reviewer: ```suggestion
Following a hunch, she adds more logging in and around the `process_work_item` function and waits a few days to accumulate new logs. She notices that shortly after each time out, there is always a log of a `process_work_item` call that takes at least 20 seconds. These calls are not related to the connections that time out, they are for other connections, but they always appear afterwards in time.

`process_work_item` is expected to be slow sometimes because it can end up handling large items, so this is not immediately surprising to Barbara. She is, however, surprised by the correlation - surely the executor ensures that `process_work_item` can't stop `do_select` from doing its job?
```

[`FuturesUnordered`]: https://docs.rs/futures/0.3.14/futures/stream/struct.FuturesUnordered.html

Barbara also knows that every future winds up associated with a task, and that if you have multiple futures on the same task (in this case, the futures in the stream, for example) then they would run concurrently, but not in parallel. Based on this, she thinks perhaps that `process_work_item` is a CPU hog that takes too long to complete, and so she needs to add a call to `spawn_blocking`. But when she looks more closely, she realizes that `process_work_item` is an async function, and those 20 seconds that it spends executing are mostly spent waiting on I/O. Huh, that's confusing, because the task ought to be able to execute other futures in that case -- so why are her connections stalling out without making progress?

### Barbara goes deep into how poll works

> She goes to read the Rust async book and tries to think about the model, but she can't quite see the problem. Then she asks on the rust-lang Discord and someone explains to her what is going on. Finally, after reading over what they wrote a few times, and reading some chapters in the async book, she sees the problem.

Reviewer: ```suggestion
She goes to read the Rust async book and tries to think about the model, but she can't quite see the problem. Then she asks on the rust-lang Discord and someone explains to her what is going on, with the catchphrase "remember, `async` is about waiting in parallel, not working in parallel". Finally, after reading over what they wrote a few times, and reading some chapters in the async book, she sees the problem.
```

When it executes `Stream::for_each`, the task is doing something like this:

```rust
while let Some(work_item) = stream.next().await {
    process_work_item(database, work_item).await;
}
```

The task can only "wait" on one "await" at a time. It will execute that await until it completes and only then move on to the rest of the function. When the task is blocked on the first `await`, it will process all the futures that are part of the stream, and hence the various buffered connections all make progress. 

But once a work item is produced, the task will block on the *second* `await` -- the one that resulted from `process_work_item`. This means that, until `process_work_item` completes, control will *never return to the first `await`*. As a result, none of the futures in the stream will make progress, even if they could do so!

### The fix

> Once Barbara understands the problem, she considers the fix. The most obvious fix is to spawn out tasks that will process each work item, like so:
> 
> ```rust
> async fn do_work(database: &Database) {
>     let work = do_select(database, FIND_WORK_QUERY)?;
>     stream::iter(work.into_iter())
>         .map(|item| do_select(database, work_from_item(item)))
>         .buffered(5)
>         .for_each(|work_item| task::spawn(async move {
>             process_work_item(database, work_item).await
>         })
>         .await;
> }
> ```

Reviewer: ```suggestion
Once Barbara understands the problem, she considers the fix. The most obvious fix is to spawn out tasks that will allow `do_select` to finish even if `process_work_item` is slow, like so:

```rust
async fn do_work(database: &Database) {
    let work = do_select(database, FIND_WORK_QUERY)?;
    stream::iter(work.into_iter())
        .map(|item| task::spawn(do_select(database, work_from_item(item))))
        .buffered(5)
        .for_each(|work_item| process_work_item(database, work_item).await)
        .await;
}
```
```

Spawning a task completes immediately, so the task will quickly move back to blocking on the stream and allowing other connections to make progress. Unfortunately, this change results in a compilation error:

```

> error[E0759]: `database` has an anonymous lifetime `'_` but it needs to satisfy a `'static` lifetime requirement
>   --> src/main.rs:8:18
>    |
> 8  | async fn do_work(database: &Database) {
>    |                  ^^^^^^^^  --------- this data with an anonymous lifetime `'_`...
>    |                  |
>    |                  ...is captured here...
>    |        .for_each(|work_item| task::spawn(async move {
>    |                              ----------- ...and is required to live as long as `'static` here
> ```

Reviewer: ```suggestion
error[E0759]: `database` has an anonymous lifetime `'_` but it needs to satisfy a `'static` lifetime requirement
  --> src/main.rs:8:18
   |
8  | async fn do_work(database: &Database) {
   |                  ^^^^^^^^  --------- this data with an anonymous lifetime `'_`...
   |                  |
   |                  ...is captured here...
   |        .map(|item| task::spawn(do_select(database, work_from_item(item))))
   |                    ----------- ...and is required to live as long as `'static` here
```
```

Reviewer: This change brings it into line with what we try as our first effort to fix this.

> "Ah, right," she says, "spawned tasks can't use borrowed data. I wish I had [rayon] or the scoped threads from [crossbeam]." (What Barbara doesn't realize is that spawning wouldn't actually have fixed her problem anyway: the `for_each` combinator would have awaited the resulting `JoinHandle` and hence it would have blocked... but she could have tweaked her program to fix that if she had gotten that far.)

Reviewer: ```suggestion
"Ah, right," she says, "spawned tasks can't use borrowed data. I wish I had [rayon] or the scoped threads from [crossbeam]."
```

> "Ah, right," she says, "spawned tasks can't use borrowed data. I wish I had [rayon] or the scoped threads from [crossbeam]." (What Barbara doesn't realize is that spawning wouldn't actually have fixed her problem anyway: the `for_each` combinator would have awaited the resulting `JoinHandle` and hence it would have blocked... but she could have tweaked her program to fix that if she had gotten that far.)

Reviewer 2: It's true that the `spawn` used here would not have fixed her problem, but spawning _would_ have fixed her problem if she had put the spawn around `do_select` rather than around `process_work_item`. (Ignoring any issues regarding `'static`)

Author: Yes, I should note that in the FAQ,

Reviewer: The fix we use most commonly is described below - we spawn inside `do_select` to ensure that the work done there can complete.

"Let me see," Barbara thinks. "What else could I do?" She has the idea that she doesn't have to process the work items immediately. She could buffer up the work into a [`FuturesUnordered`] and process it after everything is ready:

```rust
async fn do_work(database: &Database) {
    let work = do_select(database, FIND_WORK_QUERY)?;
    let mut results = FuturesUnordered::new();
    stream::iter(work.into_iter())
        .map(|item| do_select(database, work_from_item(item)))
        .buffered(5)
        .for_each(|work_item| {
            results.push(process_work_item(database, work_item));
            futures::future::ready(())
        })
        .await;

    while let Some(_) = results.next().await { }
}
```

> This isn't maximally efficient -- it's introducing an arbitrary phasing into her work -- but at least it doesn't cause timeouts. Going forward, she tries to remember never to do a m"nested `await`" like this. Buffering up work into a `FuturesUnordered` becomes a pattern she finds herself using throughout the codebase.

Reviewer 3: ```suggestion
This isn't maximally efficient -- it's introducing an arbitrary phasing into her work -- but at least it doesn't cause timeouts. Going forward, she tries to remember never to do a "nested `await`" like this. Buffering up work into a `FuturesUnordered` becomes a pattern she finds herself using throughout the codebase.
```

## 🤔 Frequently Asked Questions

### **What are the morals of the story?**

* Rust's future model is a 'leaky abstraction' that works quite differently from futures in other languages. It is prone to some subtle bugs that require a relatively deep understanding of its inner works to understand and fix.
* "Nested awaits" -- where the task blocks on an inner await while there remains other futures that are still awaiting results -- are easy to do but can cause a lot of trouble.
* Lack of scoped futures makes it hard to spawn items into separate tasks for independent processing sometimes.

### **What are the sources for this story?**

> This is based on the bug report [Footgun with Future Unordered](https://github.com/rust-lang/futures-rs/issues/2387) but the solution that Barbara came up with is something that was relayed by [farnz](https://github.com/farnz) vision doc writing session. [farnz] mentioned at the time that this pattern was frequently used in their codebase to work around this sort of hazard.

Reviewer: We don't use this pattern frequently - it's a hack to handle places where refactoring so that the `FuturesUnordered` just polls `JoinHandle`s is hard.

The normal pattern is to refactor `do_select`:

```
pub async fn do_select(...) -> Result<_, anyhow::Error> {
    // Do anything that depends on borrows from the arguments - noting that we can't `.await` at this point.
    let query = format_query!(SELECT_QUERY, ...);
    // Then spawn off a block that has 'static lifetime and does the real work, with `.await` as needed.
    tokio::spawn(async move {
        ...
    }).await?
}
```
but we use the other pattern for cases where we need to `.await` before finishing with all our borrows.

Also worth noting that the documented pattern here that I mentioned in the vision session radically changes behaviour - the solution that gets the behaviour we intended (at most 5 work handles dequeued by `do_select`, at most one work item in flight with `process_work_item`) is the spawn inside `do_select`

Author: Ah, thanks! This is super helpful! I wasn't really sure what behavior you were shooting for.

### Besides timeouts for clients, what else could go wrong?

The original bug report mentioned the possibility of deadlock:

> When using an async friendly semaphore (like Tokio provides), you can deadlock yourself by having the tasks that are waiting in the `FuturesUnordered` owning all the semaphores, while having an item in a `.for_each()` block after `buffer_unordered()` requiring a semaphore.

### Is there any way for Barbara to both produce and process work items simultaneously?

Yes, in this case, she could've. For example, she might have written

```rust
async fn do_work(database: &Database) {
    let work = do_select(database, FIND_WORK_QUERY).await?;

    stream::iter(work)
        .map(|item| async move {
            let work_item = do_select(database, work_from_item(item)).await;
            process_work_item(database, work_item).await;
        })
        .buffered(5)
        .for_each(|()| std::future::ready(()))
        .await;
}
```

> This would mean however that limit the number of work items being processed to 5. This may be what she wanted, but perhaps not. It may also be that similar problems come up in scenarios where Barbara can't so easily alter the call to `do_select` to "append" a call to `process_work_item` (for example, maybe she is in some function that takes a [`FuturesUnordered`] or `Stream` as argument).

Reviewer: Note that it's common for us to have cases where `do_select` consumes *maybe* a kilobyte of RAM peak (as it returns a handle to work on), while `process_work_item` can consume 5 to 10 GiB peak, as it grabs compressed datablobs from an S3-like storage system, decompresses them, parses them, converts to another format, then sends them over TCP to a different service.

We want `do_select` to be allowed to run ahead to hide the latency of a database fetch, but `process_work_item` can't be allowed to run too parallel because it would consume too much RAM if it did.

### Is there any way for Barbara to both produce and process work items simultaneously, without the buffering and so forth?

Yes, she might use a loop with a `select!`. This would ensure that she is processing *both* the stream that produces work items and the [`FuturesUnordered`] that consumes them:

```rust
async fn do_work(database: &Database) {
    let work = do_select(database, FIND_WORK_QUERY).await?;

    let selects = stream::iter(work)
        .map(|item| do_select(database, work_from_item(item)))
        .buffered(5)
        .fuse();
    tokio::pin!(selects);

    let mut results = FuturesUnordered::new();

    loop {
        tokio::select! {
            Some(work_item) = selects.next() => {
                results.push(process_work_item(database, work_item));
            },
            Some(()) = results.next() => { /* do nothing */ },
            else => break,
        }
    }
}
```

> Note that doing so is producing code that looks quite a bit different than where she started, though. :(

Reviewer: This also behaves very differently - we can have a queue of tens of thousands of items that `do_select` grabs from, and we don't want to pull too many items out of the queue (because then we'd have to re-queue them on shutdown).

The intent of `buffered(5)` here was to allow us to grab 5 work items from the queue at most, so that other hosts could pull out work items and share the load when there's a spike. It doesn't work like that, unfortunately :-(

Author: I incorporated this
